wagtail_streamfield_migration_toolkit/autodetect/comparers.py

```diff
@@ -1,156 +1,269 @@
+    # weights for the importance of argument similarity and name similarity
     arg_weight = None
-    child_weight = None
-    # name_weight = None # TODO
+    name_weight = None
+    kwarg_weight = None
 
     @classmethod
-    def compare(cls, old_def, old_name, new_def, new_name):
+    @lru_cache()
+    def _compare(
+        cls,
+        old_name,
+        old_path,
+        old_args,
+        old_kwargs,
+        new_name,
+        new_path,
+        new_args,
+        new_kwargs,
+    ):
+        # This might not be a black and white decision for all cases. However for cases like where
+        # old block is a StreamBlock and compared block is a CharBlock, it is fair to say that we
+        # can be certain it is not a renamed block.
+        # TODO consider eg. like CharBlock, TextBlock. In that case do we give this a weight and
+        # return a non zero score for blocks like char and text that may have compatible content?
+        if not cls.compare_types(old_path, new_path):
+            return 0
+
+        name_similarity = cls.compare_names(old_name, new_name)
+        arg_similarity = cls.compare_args(old_args, new_args)
+        kwarg_similarity = cls.compare_kwargs(old_kwargs, new_kwargs)
 
-        # TODO this might not be a black and white decision for all cases
-        # However for cases like where old block is a StreamBlock and compared block is a CharBlock,
-        # it is fair to say that we can be certain it is not a renamed block.
-        # TODO consider eg. like CharBlock, TextBlock
-        if not cls.compare_types(old_def, new_def):
-            return False
-
-        # name_similarity = cls.compare_names(old_name, new_name) # TODO
+        return (
+            arg_similarity * cls.arg_weight
+            + name_similarity * cls.name_weight
+            + kwarg_similarity * cls.kwarg_weight
+        ) / (cls.arg_weight + cls.name_weight + cls.kwarg_weight)
 
-        arg_similarity = cls.compare_args(old_def, new_def)
+    @classmethod
+    def compare(cls, old_def, old_name, new_def, new_name):
 
-        child_similarity = cls.compare_children(old_def, new_def)
+        if not cls.compare_types_initial(old_def, new_def):
+            return 0
+
+        # TODO it might be best to add some separate tests for the hashable_deep_deconstruct method
+        # itself
+        old_path, old_args, old_kwargs = cls.hashable_deep_deconstruct(old_def)
+        new_path, new_args, new_kwargs = cls.hashable_deep_deconstruct(new_def)
+
+        # - For structural blocks, args are a list of children, and kwargs contain block options
+        # like label, icon etc.
+        # - For other blocks, args includes any positional arguments and kwargs contains the block
+        # options. For example, SnippetChooser block would have the snippet class as a positional
+        # argument. Most basic blocks like CharBlock have no args it seems.
```

Comment:
Oops, it seems like that would be a problem for some blocks like `SnippetChooserBlock`. It does always hold for structural blocks since they have a deconstruct method which sets the children as args.